src/utils/helpers/DrawHelpers.ts

```diff
@@ -0,0 +1,223 @@
+                .on('zoom', zoomed)
+        )
+        .on('dblclick.zoom', null);
+
+    return zoomLayer;
+}
+
+export function createLinkPath(this: Element, link: DependencyLink): void {
+    if (!link.source.x || !link.source.y || !link.target.x || !link.target.y) {
+        return;
+    }
+
+    const xDiff = link.source.x - link.target.x;
+    const yDiff = link.source.y - link.target.y;
+
+    const isSourceOnTheLeft = xDiff < 0;
+    const isSourceBelowTarget = yDiff > 0;
+
+    const angleInRadians = Math.abs(Math.atan(yDiff / xDiff));
+    const cosinus = Math.cos(angleInRadians);
+    const sinus = Math.sin(angleInRadians);
+
+    const offsetXLeft = -50 * cosinus;
+    const offsetY = 50 * sinus;
+    const offsetYBelow = -offsetY - 5;
+
+    const sourceLabelWidth = getNodeDimensions(link.source).width;
+    link.source.width = sourceLabelWidth;
+    const targetLabelWidth = getNodeDimensions(link.target).width;
+    link.target.width = targetLabelWidth;
+
+    const sourceNewX = isSourceOnTheLeft ? (sourceLabelWidth + 20) * cosinus : offsetXLeft;
+    const sourceNewY = isSourceBelowTarget ? offsetYBelow : offsetY;
+
+    const targetNewX = isSourceOnTheLeft ? offsetXLeft : (targetLabelWidth + 20) * cosinus;
+    const targetNewY = isSourceBelowTarget ? offsetY : offsetYBelow;
+
+    select<Element, DependencyLink>(this)
+        .attr(
+            'd',
+            'M' +
+                (link.source.x + sourceNewX) +
+                ',' +
+                (link.source.y + sourceNewY) +
+                ', ' +
+                (link.target.x + targetNewX) +
+                ',' +
+                (link.target.y + targetNewY)
+        )
+        .attr('stroke', Colors.DEFAULT);
+}
+
+export function createLabelPath(this: Node, node: DependencyNode) {
+    const labelTextDimensions = getLabelTextDimensions(this);
+
+    if (!labelTextDimensions) {
+        return '';
+    }
+
+    const labelTextWidth = labelTextDimensions.width;
+
+    const { isConsumer, isProvider } = node;
+
+    if (isConsumer && isProvider) {
+        return 'M4.5,35l9.37,14.59L4.5,64.18h' + (labelTextWidth + 45) + 'l9-14.59L' + (labelTextWidth + 49.5) + ',35H4.5z';
+    }
+
+    if (isProvider) {
+        return 'M' + (labelTextWidth + 49.5) + ',35H4.5l9.37,14.59L4.5,64.18h' + (labelTextWidth + 45);
+    }
+
+    if (isConsumer) {
+        return 'M4.5,64.18h' + (labelTextWidth + 45) + 'l9.42-14.59L' + (labelTextWidth + 49.5) + ',35H4.5';
+    }
+
+    return 'M4.5,64.18h' + (labelTextWidth + 55) + 'L' + (labelTextWidth + 59.5) + ',35H4.5';
+}
+
+export function setKeyboardDependencyHighLightHandler() {
+    select('body').on('keydown', () => {
+        const labelNodesGroup = select<SVGGElement, DependencyNode>('g#labels');
+        let maxLevel = findMaxDependencyLevel(labelNodesGroup);
+
+        if (!isFinite(maxLevel)) {
+            return;
+        }
+
+        if (LevelStorage.getLevel() < maxLevel - 1 && event.code === 'NumpadAdd') {
```

Comment:
No proper **max level value** for ``maxLevel``.

Max value for ``LevelStorage.level`` is ``maxLevel - 1``.  So ``maxLevel`` is not **max level**, but **max level - 1**.

Reply:
Good point.